core/src/string/interner.rs

```diff
@@ -91,14 +113,49 @@ impl<'gc> AvmStringInterner<'gc> {
         let atom = match self.interned.entry(mc, s.as_wstr()) {
             (Some(atom), _) => atom,
             (None, h) => {
-                let repr = s.to_owned(mc);
+                let repr = s.to_fully_owned(mc);
                 repr.mark_interned();
                 self.interned.insert_fresh(mc, h, repr)
             }
         };
 
         AvmAtom(atom)
     }
+
+    #[must_use]
+    pub fn empty(&self) -> AvmString<'gc> {
+        self.empty.into()
+    }
+
+    #[must_use]
+    pub fn get_char(&self, mc: &Mutation<'gc>, c: u16) -> AvmString<'gc> {
+        if c < 128 {
+            self.chars[c as usize].into()
+        } else {
+            AvmString::new(mc, WString::from_unit(c))
+        }
+    }
+
+    #[must_use]
+    pub fn substring(
+        &self,
+        mc: &Mutation<'gc>,
+        s: AvmString<'gc>,
+        start_index: usize,
+        end_index: usize,
+    ) -> AvmString<'gc> {
+        if start_index >= end_index {
```

Comment:
IMHO, this should panic if `start_index > end_index`, for consistency with slicing syntax (and maybe it should even take a `range: impl RangeBounds<usize>`?)